src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs

```diff
@@ -38,117 +38,109 @@ private static class Compiler<TModel, TResult>
+            }
 
-                var memberExpression = expression.Body as MemberExpression;
-                if (memberExpression != null)
+            private static Func<TModel, TResult> CompileFromStaticMemberAccess(
+                Expression<Func<TModel, TResult>> expression,
+                MemberExpression memberExpression)
+            {
+                // model => model.StaticMember
+                if (_simpleMemberAccessCache.TryGetValue(memberExpression.Member, out var result))
                 {
-                    if (memberExpression.Expression == expression.Parameters[0] || memberExpression.Expression == null)
-                    {
-                        // model => model.Member or model => StaticMember
-                        return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ => expression.Compile());
-                    }
-
-                    var constantExpression = memberExpression.Expression as ConstantExpression;
-                    if (constantExpression != null)
-                    {
-                        // model => {const}.Member (captured local variable)
-                        var compiledExpression = _constMemberAccessCache.GetOrAdd(memberExpression.Member, _ =>
-                        {
-                            // rewrite as capturedLocal => ((TDeclaringType)capturedLocal).Member
-                            var parameterExpression = Expression.Parameter(typeof(object), "capturedLocal");
-                            var castExpression =
-                                Expression.Convert(parameterExpression, memberExpression.Member.DeclaringType);
-                            var replacementMemberExpression = memberExpression.Update(castExpression);
-                            var replacementExpression = Expression.Lambda<Func<object, TResult>>(
-                                replacementMemberExpression,
-                                parameterExpression);
-
-                            return replacementExpression.Compile();
-                        });
-
-                        var capturedLocal = constantExpression.Value;
-                        return _ => compiledExpression(capturedLocal);
-                    }
+                    return result;
                 }
 
-                return null;
+                result = expression.Compile();
+                result = _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, result);
+                return result;
             }
 
-            private static Func<TModel, TResult> CompileSlow(Expression<Func<TModel, TResult>> expression)
+            private static Func<TModel, TResult> CompileCapturedConstant(MemberExpression memberExpression, ConstantExpression constantExpression)
             {
-                if (expression == null)
+                // model => {const}.Member (captured local variable)
+                if (!_constMemberAccessCache.TryGetValue(memberExpression.Member, out var result))
```

Comment:
Again, why add this short-circuit?

If this is a performance measure, have you verified the extra complexity matters? If yes, add a comment indicating the benefit e.g. avoids creating the `_ => { ... }` delegate.

Reply:
If this remains, use the same short-circuiting pattern as in `CompileFrom[Static]MemberAccess(...)`.

Reply:
Are you saying why use this versus GetOrAdd? I think we more commonly use the `TryGetValue` \ `GetOrAdd` pattern - 

* https://github.com/aspnet/Mvc/blob/79c4986685febe0cbb03a50019cacc99598d7aa2/src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerCache.cs#L69
* https://github.com/aspnet/Mvc/blob/b7db6dbc8e75c6c9b02daed1e841027cd14b6450/src/Microsoft.AspNetCore.Mvc.Razor/RazorPageActivator.cs#L67
* https://github.com/aspnet/Mvc/blob/599d1a502ef358678202345ced2d11ddca0754f3/src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvokerProvider.cs#L105

etc

It's just more like the rest of our code.

Reply:
> It's just more like the rest of our code.

It isn't -- we use `GetOrAdd(...)` alone a great deal. However, I can buy into not creating a (capturing) lambda per expression per view per request.

Leave the `TryGetValue(...)` additions in.

> Doesn't work the same as the other one.

Then please update the other one to match this approach. That is, handle the caches consistently in this class and avoid future @pranavkm wondering why methods differ.

Reply:
They cache different things. One caches the result, and hence can be returned, the other caches the intermediate result.  Leaving this as is, because it makes sense with a cursory glance of the code

Reply:
Ugh but :ok: